fpga_interchange/yosys_json.py

```diff
@@ -0,0 +1,417 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2021  The SymbiFlow Authors.
+#
+# Use of this source code is governed by a ISC-style
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/ISC
+#
+# SPDX-License-Identifier: ISC
+""" Utility for converting yosys json to logical netlist. """
+import argparse
+import json
+
+from fpga_interchange.interchange_capnp import Interchange, write_capnp_file
+from fpga_interchange.logical_netlist import LogicalNetlist, Cell, \
+        CellInstance, Direction, Library
+
+
+def is_bus(bits, offset, upto):
+    """ Returns true if this port/net is a bus.
+
+    >>> is_bus([2], offset=0, upto=0)
+    False
+    >>> is_bus([2, 3], offset=0, upto=0)
+    True
+    >>> is_bus([2], offset=1, upto=0)
+    True
+    >>> is_bus([2], offset=0, upto=1)
+    True
+
+    """
+    return len(bits) > 1 or offset != 0 or upto != 0
+
+
+def interp_yosys_net(bits, offset, upto):
+    """ Convert yosys json bits/offset/upto into (net selection, net bit)
+
+    Yosys JSON encodes multi-bit nets as a triple of bits / offset / upto:
+     - bits is a list of net bits that uniquely identifies what net the
+       signal is tied too.
+     - offset (default: 0) is the net offset.  Examples:
+
+         wire [7:4] ex1; // offset = 4
+         wire [3:0] ex2; // offset = 0 (implicitly)
+         wire [0:3] ex3; // offset = 0 (implicitly)
+         wire [5:7] ex4; // offset = 5
+
+    - upto (default: 0) determines if the net is count up or count down.
+
+         wire [7:4] ex1; // upto = 0 (implicitly)
+         wire [3:0] ex2; // upto = 0 (implicitly)
+         wire [0:3] ex3; // upto = 1
+         wire [5:7] ex4; // upto = 1
+
+    >>> bits = [2, 3, 4, 5]
+    >>> tuple(interp_yosys_net(bits, offset=0, upto=0))
+    ((0, 2), (1, 3), (2, 4), (3, 5))
+    >>> tuple(interp_yosys_net(bits, offset=4, upto=0))
+    ((4, 2), (5, 3), (6, 4), (7, 5))
+    >>> tuple(interp_yosys_net(bits, offset=0, upto=1))
+    ((3, 2), (2, 3), (1, 4), (0, 5))
+    >>> tuple(interp_yosys_net(bits, offset=4, upto=1))
+    ((7, 2), (6, 3), (5, 4), (4, 5))
+
+    """
+    if not upto:
+        for idx, bit in enumerate(bits):
+            yield idx + offset, bit
+    else:
+        for idx, bit in enumerate(bits):
+            yield offset + (len(bits) - 1) - idx, bit
+
+
+# FIXME: These constants should likely be from the device database.
```

Comment:
Admittedly I don't have much context here, but have you thought about having constants as a special case in the schema? Requiring a netlist transformation to know about device-specific cells seems a bit suboptimal to me.

Reply:
This is a gap that needs to be filled, both for this and the pin map case.